sdks/java/harness/src/test/java/org/apache/beam/fn/harness/state/StateBackedIterableTest.java

```diff
@@ -0,0 +1,184 @@
+import org.junit.runners.JUnit4;
+import org.junit.runners.Parameterized;
+
+/** Tests for {@link StateBackedIterable}. */
+@RunWith(Enclosed.class)
+public class StateBackedIterableTest {
+
+  @RunWith(Parameterized.class)
+  public static class IterationTest {
+    @Parameterized.Parameters
+    public static Iterable<Object[]> data() {
+      return ImmutableList.<Object[]>builder()
+          .add(new Object[] {Collections.emptyList(), "emptySuffix", ImmutableList.of()})
+          .add(new Object[] {ImmutableList.of("A", "B"), "emptySuffix", ImmutableList.of("A", "B")})
+          .add(
+              new Object[] {
+                Collections.emptyList(),
+                "nonEmptySuffix",
+                ImmutableList.of("C", "D", "E", "F", "G", "H", "I", "J", "K")
+              })
+          .add(
+              new Object[] {
+                ImmutableList.of("A", "B"),
+                "nonEmptySuffix",
+                ImmutableList.of("A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K")
+              })
+          .build();
+    }
+
+    @Parameterized.Parameter(0)
+    public List<String> prefix;
+
+    @Parameterized.Parameter(1)
+    public String suffixKey;
+
+    @Parameterized.Parameter(2)
+    public List<String> expected;
+
+    @Test
+    public void testIteration() throws Exception {
+      FakeBeamFnStateClient fakeBeamFnStateClient =
+          new FakeBeamFnStateClient(
+              ImmutableMap.of(
+                  key("nonEmptySuffix"), encode("C", "D", "E", "F", "G", "H", "I", "J", "K"),
+                  key("emptySuffix"), encode()));
+
+      StateBackedIterable<String> iterable =
+          new StateBackedIterable<>(
+              fakeBeamFnStateClient,
+              "instruction",
+              encode(suffixKey),
+              StringUtf8Coder.of(),
+              prefix);
+
+      assertEquals(expected, Lists.newArrayList(iterable));
+      assertEquals(expected, Lists.newArrayList(iterable));
+      assertEquals(expected, Lists.newArrayList(iterable));
+    }
+
+    @Test
+    public void testUsingMultipleInterleavedIterators() throws Exception {
+      FakeBeamFnStateClient fakeBeamFnStateClient =
+          new FakeBeamFnStateClient(
+              ImmutableMap.of(
+                  key("nonEmptySuffix"), encode("C", "D", "E", "F", "G", "H", "I", "J", "K"),
+                  key("emptySuffix"), encode()));
+
+      StateBackedIterable<String> iterable =
+          new StateBackedIterable<>(
+              fakeBeamFnStateClient,
+              "instruction",
+              encode(suffixKey),
+              StringUtf8Coder.of(),
+              prefix);
+
+      List<Iterator<String>> iterators = new ArrayList<>();
+      List<List<String>> results = new ArrayList<>();
+      for (int i = 0; i < 3; ++i) {
+        iterators.add(iterable.iterator());
+        results.add(new ArrayList<>());
+      }
+
+      Random random = new Random(42L);
+      while (!iterators.isEmpty()) {
```

Comment:
Does it make sense to add a test case for re-iteration?

Reply:
This is covering reiteration since the same iterable is producing multiple iterators that are read in random order.

`testIteration` also covers re-iteration, renamed test for clarity